save-cosv/src/main/kotlin/com/saveourtool/save/cosv/controllers/CosvController.kt

```diff
@@ -1,88 +1,205 @@
+import com.saveourtool.save.permission.Permission
 import com.saveourtool.save.utils.*
 import com.saveourtool.save.v1
-
+import org.springframework.http.HttpHeaders
+import org.springframework.http.HttpStatus
 import org.springframework.http.MediaType
 import org.springframework.http.ResponseEntity
 import org.springframework.http.codec.multipart.FilePart
 import org.springframework.security.core.Authentication
 import org.springframework.web.bind.annotation.*
 import reactor.core.publisher.Flux
 import reactor.core.publisher.Mono
-
-import kotlinx.serialization.encodeToString
-import kotlinx.serialization.json.Json
+import reactor.core.scheduler.Schedulers
 
 /**
  * Rest controller for COSVs
  */
 @ApiSwaggerSupport
 @RestController
-@RequestMapping("/api/$v1/cosv")
+@RequestMapping("/api/$v1/cosv/{organizationName}")
 class CosvController(
     private val cosvService: CosvService,
+    private val rawCosvFileStorage: RawCosvFileStorage,
+    private val backendService: IBackendService,
 ) {
     /**
-     * @param cosvId COSV identifier
-     * @return content of COSV
+     * @param organizationName
+     * @param content
+     * @param authentication
+     * @return uploaded [RawCosvFileDto]
      */
     @RequiresAuthorizationSourceHeader
-    @GetMapping(path = ["/get-by-id/{cosvId}"], produces = [MediaType.APPLICATION_JSON_VALUE])
-    fun getById(
-        @PathVariable cosvId: String,
-    ): Mono<StringResponse> = cosvService.findById(cosvId)
-        .map {
-            ResponseEntity.ok(Json.encodeToString(it))
+    @PostMapping("/upload", consumes = [MediaType.APPLICATION_JSON_VALUE])
+    fun upload(
+        @PathVariable organizationName: String,
+        @RequestBody content: String,
+        authentication: Authentication,
+    ): Mono<RawCosvFileDto> = hasPermission(authentication, organizationName, Permission.WRITE, "upload")
+        .flatMap {
+            rawCosvFileStorage.upload(
+                key = RawCosvFileDto(
+                    "manually_uploaded_${getCurrentLocalDateTime()}.json",
+                    organizationName = organizationName,
+                    userName = authentication.name,
+                ),
+                contentBytes = content.toByteArray(),
+            )
         }
 
     /**
-     * @param content
+     * @param organizationName
+     * @param filePartFlux
      * @param authentication
+     * @return list of uploaded [RawCosvFileDto]
+     */
+    @RequiresAuthorizationSourceHeader
+    @PostMapping("/batch-upload", consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
+    fun batchUpload(
+        @PathVariable organizationName: String,
+        @RequestPart(FILE_PART_NAME) filePartFlux: Flux<FilePart>,
+        authentication: Authentication,
+    ): Flux<RawCosvFileDto> = hasPermission(authentication, organizationName, Permission.WRITE, "upload")
+        .flatMapMany {
+            filePartFlux
+                .flatMap { filePart ->
```

Comment:
may be parallel() here? Looks like these files are very independently processed

Reply:
will add

Reply:
added